### PR TITLE
Release/v4.0.0

### PR DESCRIPTION
**Pull request by creisle**
Breaking Changes
- No longer using classes for ClassDefinition and PropertyDefinition, see migration guide in the README. 

https://github.com/bcgsc/pori_graphkb_schema/tree/release/v4.0.0#migrating-from-v3-to-v4

### CODE REVIEW

**Review by areisle — APPROVED** [review]

**Comment by areisle** [comment]
the `validateProperty` function will need to be exported, as the gui uses `Property.validateWith` atm and doesn't know the modelName or propName at that point in the code

I've also noticed some issues with the migration guide. validate is missing from it, and the types for `queryProperties` say they return arrays but seem to be actually objects

File: README.md [hunk 1]
```diff
@@ -51,18 +51,17 @@ constants.RID = RID; // IMPORTANT: Without this all castToRID will do is convert
...
+| `ClassModel.inherits`                                    | `SchemaDefinition.ancestors(modelName: string)`                                                                        |                                                          |
+| `ClassModel.subclasses: ClassModel[]`                    | `SchemaDefinition.children(modelName: string): string[]`                                                               |                                                          |
+| `ClassModel.descendantTree(boolean): ClassModel[]`       | `SchemaDefinition.descendants(modelName: string, opt: { excludeAbstract?: boolean, includeSelf?: boolean }): string[]` | must be called with includeSelf=true to match v3 edition |
+| `ClassModel.queryProperties: Property[]`                 | `SchemaDefinition.queryableProperties(modelName: string): Record<string,PropertyDefinition[]>`                         |                                                          |
```
Comment (areisle) [inline README.md:63]: this is still incorrect. it was never an array

File: README.md [hunk 2]
```diff
@@ -51,18 +51,17 @@ constants.RID = RID; // IMPORTANT: Without this all castToRID will do is convert
...
-| `formatRecord(record: GraphRecord, opt = {})` | `formatRecord(modelName: string, record: GraphRecord, opt = {})`                                      |                                                          |
-
-PropertyDefinition objects also now store the class name under linkedClass as a string rather than a reference to the actual ClassDefinition instance
+| v3                                                       | v4 equivalent                                                                                                          | Notes                                                    |
+| -------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------- |
+| `ClassModel.properties`                                  | `SchemaDefinition.getProperties(modelName: string)`                                                                    |                                                          |
```
Comment (areisle) [inline README.md:56]: it seems like `ClassDefinition.properties` is still a thing, is that different than using `SchemaDefinition.getProperties(modelName: string)`?

Comment (creisle) [reply README.md:56]: I will add a note for this

File: README.md (hunk 1 above)
Comment (creisle) [reply README.md]: fixed

File: README.md (hunk 2 above)
Comment (creisle) [reply README.md]: https://github.com/bcgsc/pori_graphkb_schema/pull/20/commits/a710e41f1b6bc05522535b5508461941e6185769

**Comment by areisle** [comment]
I've been working on migrating the gui using this branch, and the gui ends up copying and then modifying a lot of properties, but the only type `PropertyDefinition` that exported is entirely readonly so I end up with a lot of typescript errors typing to use it. could this be changed to not have the readonly modifiers so I can use it in the gui and then where you want to use it as readonly in the schema package wrap it in ReadOnly (`ReadOnly<PropertyDefinition>`)?

File: test/schema.test.ts [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+    ['blargh', { displayName: 'blargh', name: 'monkeys' }],
+    ['blargh', { name: 'blargh' }],
+    ['DNMT3A:p.R882 predicts unfavourable prognosis in acute myeloid leukemia [DOID:9119]', { ...examples['subject:null|conditions:Disease;PositionalVariant|relevance:unfavourable prognosis'], '@class': 'Statement' }],
+    ['#3:4', { '@rid': '#3:4' }],
+    ['3', [1, 2, 3]],
+    ['blargh', { target: { name: 'blargh' } }],
+    ['blargh', 'blargh']
+])('getPreview %s', (preview, input) => {
+    // @ts-ignore  testing bad input in some cases
+    expect(schema.getPreview(input)).not.toHaveProperty(preview);
```
Comment (areisle) [inline test/schema.test.ts:377]: noticed when migrating the gui that getPreview was not behaving as expected.
I missed this in PRs to develop, but I think this must have been a copy paste error and should actually be
```ts
expect(schema.getPreview(input)).toEqual(preview);
```

additionally would like these cases from the gui tests to be added
```js
 [
  'User',
  {
      '@rid': '20:0',
      '@class': 'User',
      createdBy: 'Mom',
      deletedBy: 'Mom',
  },
],
[
  '22:0',
  {
      '@rid': '22:0',
  },
],
[
  'Disease',
  {
      '@class': 'disease',
  },
],
[
  'Given Low blood sugar Mood Swings applies to hungertitis (A reputable source)',
  {
      displayName: 'displayName',
      '@class': 'Statement',
      '@rid': '22:0',
      displayNameTemplate: 'Given {conditions} {relevance} applies to {subject} ({evidence})',
      relevance: { displayName: 'Mood Swings' },
      conditions: [{ displayName: 'Low blood sugar' }],
      subject: { displayName: 'hungertitis' },
      evidence: [{ displayName: 'A reputable source' }],
  },
],
[
  '#19:0', '#19:0',
],
```

**Review by areisle — CHANGES_REQUESTED** [review]

File: test/schema.test.ts (hunk 3 above)
Comment (creisle) [reply test/schema.test.ts:377]: I've fixed some of these cases, but the requirement of `@rid` in the statement was on purpose. There should never be a case where the elements linked to the statement don't have an `@rid`. Are there any cases where that's not true in the current client implementation?

Comment (areisle) [reply test/schema.test.ts:377]: I couldn't say. it's pretty hard to what shape the records are in by the time `getPreview` is called in the gui. the above cases all came from tests in the gui so I'm not sure if rid was left out on purpose or not

**Comment by creisle** [comment]
> I've been working on migrating the gui using this branch, and the gui ends up copying and then modifying a lot of properties, but the only type `PropertyDefinition` that exported is entirely readonly so I end up with a lot of typescript errors typing to use it. could this be changed to not have the readonly modifiers so I can use it in the gui and then where you want to use it as readonly in the schema package wrap it in ReadOnly (`ReadOnly<PropertyDefinition>`)?

Changed to top-level Readonly wrapper in https://github.com/bcgsc/pori_graphkb_schema/pull/20/commits/480a2265345e38e378b310247765fd928c88a52c

File: test/schema.test.ts (hunk 3 above)
Comment (areisle) [reply test/schema.test.ts]: the only case I could see for not having rid required is if you wanted to call it on a record that hasn't been created yet. I tried to look through the gui code to see if this is the cause anywhere, but couldn't tell for sure.

**Review by areisle — APPROVED** [review]
I ran into too many issues trying to migrate to this version in the gui and have decided to tidy up that code first to hopefully make things easier. I can't say for sure if they'll be other changes that need to happen in order to migrate successfully or not so I can approving this, but I could see a v5 possibly being necessary before the gui upgrades versions

File: test/schema.test.ts (hunk 3 above)
Comment (creisle) [reply test/schema.test.ts]: tests added here: https://github.com/bcgsc/pori_graphkb_schema/pull/20/commits/0c9873bb351d152dfdf0d7ef37a0c177dad121d7

**Review by areisle — APPROVED** [review]

**Review by elewis2 — APPROVED** [review]

**Pull request merged**